<a href="https://colab.research.google.com/github/Amyerm/ClassFiles/blob/main/Actividad_Calificada_07_Datos_Temporales_MATRICULA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | Ari Daniel Mendoza Enrriquez |
| **Matrícula** | 271756 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [13]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [14]:
# Primero revisar como se leyeron las dos columnas de fechas.
print('Tipos antes de convertir:')
print(taxis[['pickup', 'dropoff']].dtypes)
faltantes_antes = taxis[['pickup', 'dropoff']].isna().sum()

# Indicar el formato de año-mes-día y hora:minuto:segundo. Si una fecha
# no se pudiera convertir, errors='coerce' hace verla como NaT.
for columna in ['pickup', 'dropoff']:
    taxis[columna] = pd.to_datetime(taxis[columna], format='%Y-%m-%d %H:%M:%S', errors='coerce')

print('\nTipos después de convertir:')
print(taxis[['pickup', 'dropoff']].dtypes)
print('\nFaltantes antes y después:')
print(pd.DataFrame({'antes': faltantes_antes,
                    'después': taxis[['pickup', 'dropoff']].isna().sum()}))

Tipos antes de convertir:
pickup     object
dropoff    object
dtype: object

Tipos después de convertir:
pickup     datetime64[ns]
dropoff    datetime64[ns]
dtype: object

Faltantes antes y después:
         antes  después
pickup       0        0
dropoff      0        0


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [15]:
# Uso el inicio del 1 de abril como límite superior sin incluirlo.
inicio_marzo = pd.Timestamp('2019-03-01')
inicio_abril = pd.Timestamp('2019-04-01')
pickup_en_marzo = taxis['pickup'].between(inicio_marzo, inicio_abril, inclusive='left')
dropoff_en_marzo = taxis['dropoff'].between(inicio_marzo, inicio_abril, inclusive='left')

print('Rango de pickup:', taxis['pickup'].min(), 'a', taxis['pickup'].max())
print('Rango de dropoff:', taxis['dropoff'].min(), 'a', taxis['dropoff'].max())
print('\nViajes que iniciaron fuera de marzo:', (~pickup_en_marzo).sum())
print(taxis.loc[~pickup_en_marzo, ['pickup', 'dropoff']].to_string(index=False))
print('\nIniciaron en marzo y terminaron fuera:',
      (pickup_en_marzo & ~dropoff_en_marzo).sum())
print(taxis.loc[pickup_en_marzo & ~dropoff_en_marzo,
                ['pickup', 'dropoff']].to_string(index=False))

Rango de pickup: 2019-02-28 23:29:03 a 2019-03-31 23:43:45
Rango de dropoff: 2019-02-28 23:32:35 a 2019-04-01 00:13:58

Viajes que iniciaron fuera de marzo: 1
             pickup             dropoff
2019-02-28 23:29:03 2019-02-28 23:32:35

Iniciaron en marzo y terminaron fuera: 1
             pickup             dropoff
2019-03-31 23:43:45 2019-04-01 00:13:58


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [16]:
# Revisar las horas de inicio del domingo en que cambió el reloj.
dia_cambio = pd.Timestamp('2019-03-10').date()
viajes_cambio = taxis.loc[taxis['pickup'].dt.date == dia_cambio]
conteo_horas = viajes_cambio.groupby(viajes_cambio['pickup'].dt.hour).size()
print('Viajes que iniciaron el 10 de marzo:', len(viajes_cambio))
print('\nViajes por hora:')
print(conteo_horas.to_string())
print('Viajes iniciados entre las 02:00 y las 02:59:', conteo_horas.get(2, 0))

# Marcar como NaT las horas que no existieron en Nueva York.
pickup_ny = taxis['pickup'].dt.tz_localize('America/New_York',
                                         nonexistent='NaT', ambiguous='NaT')
print('Valores NaT al localizar pickup:', pickup_ny.isna().sum())

Viajes que iniciaron el 10 de marzo: 185

Viajes por hora:
pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
Viajes iniciados entre las 02:00 y las 02:59: 0
Valores NaT al localizar pickup: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

La conversión de las 6,433 filas no produjo NaT en pickup ni en dropoff. Encontre que un viaje que inicia el 28 de febrero, fuera del periodo de marzo. También hay uno que inicia el 31 de marzo y termina el 1 de abril, lo considero un viaje válido de marzo porque el periodo usa la fecha de inicio.
El 10 de marzo hubo 185 inicios y ninguno entre las 02:00 y las 02:59. Al localizar pickup en America/New_York con nonexistent='NaT', quedaron 0 valores NaT, esto es compatible con que las horas estén registradas en la hora local de Nueva York, aunque la ausencia de viajes a las 02:00 por sí sola no demuestra la zona horaria. No encontré una hora inexistente que necesite corregirse.

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [17]:
# La resta de dos fechas produce una duración, convertir sus segundos a minutos.
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60
print(taxis['duracion_min'].describe())

count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [18]:
# Cada máscara busca un tipo de duración sospechosa.
duracion_negativa = taxis['duracion_min'] < 0
duracion_cero = taxis['duracion_min'] == 0
duracion_corta = taxis['duracion_min'] < 1
duracion_larga = taxis['duracion_min'] > 180

print('Duración negativa:', duracion_negativa.sum())
print('Duración cero:', duracion_cero.sum())
print('Menor a un minuto (incluye las de cero):', duracion_corta.sum())
print('Mayor a tres horas:', duracion_larga.sum())
print('\nViajes con duración cero:')
print(taxis.loc[duracion_cero,
                ['pickup', 'dropoff', 'distance', 'fare', 'total']].to_string(index=False))

# Mirar distancia y tarifa para decidir qué significan los viajes cortos.
print('\nViajes cortos con distancia positiva:',
      (duracion_corta & (taxis['distance'] > 0)).sum())
print('Viajes cortos con tarifa positiva:',
      (duracion_corta & (taxis['fare'] > 0)).sum())

Duración negativa: 0
Duración cero: 6
Menor a un minuto (incluye las de cero): 72
Mayor a tres horas: 0

Viajes con duración cero:
             pickup             dropoff  distance  fare  total
2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8

Viajes cortos con distancia positiva: 32
Viajes cortos con tarifa positiva: 72


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [19]:
# Dividir las millas entre el tiempo en horas. Los viajes de duración
# cero no tienen una velocidad calculable, así que dejo NaN en ellos.
taxis['velocidad_mph'] = (taxis['distance'] /
                          (taxis['duracion_min'] / 60)).where(taxis['duracion_min'] > 0)
velocidad_alta = taxis['velocidad_mph'] > 60
print('Viajes con velocidad mayor a 60 mph:', velocidad_alta.sum())
print('De ellos, con duración menor a un minuto:',
      (velocidad_alta & duracion_corta).sum())
print('\nCinco mayores velocidades:')
columnas = ['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']
print(taxis.sort_values('velocidad_mph', ascending=False)[columnas]
      .head(5).to_string(index=False))

Viajes con velocidad mayor a 60 mph: 11
De ellos, con duración menor a un minuto: 10

Cinco mayores velocidades:
             pickup             dropoff  duracion_min  distance  velocidad_mph  fare
2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40    4834.285714   2.5
2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10    2020.000000  12.0
2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60     960.000000   2.5
2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70     680.000000   2.5
2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34     612.000000  65.0


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

No podria asumir que los 72 viajes son todos errores: 32 tienen distancia positiva y los 72 tienen una tarifa positiva, podrían ser viajes muy cortos o registros de cobro particulares. Pero para analizar duración y velocidad, una duración tan corta hace poco confiable la velocidad calculada.
Excluire esos 72 (incluidos los 6 de duración cero), tambien el viaje que inició fuera de marzo, las duraciones mayores de 180 minutos y las velocidades mayores de 60 mph. También excluiría duraciones negativas con el límite de menos de un minuto, aunque aquí hubo 0.
El límite de 180 minutos separaría recorridos inusualmente largos, pero en esta muestra no excluye ninguno (máximo: 107.67 minutos). Encontré 11 velocidades mayores de 60 mph, 10 ya están entre los viajes de menos de un minuto, por lo que no debo sumarlas dos veces. Conservo el viaje que termina el 1 de abril porque inició en marzo.

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [20]:
# Extraer las partes de la fecha de inicio para agrupar los viajes.
taxis['hora'] = taxis['pickup'].dt.hour
taxis['dia_num'] = taxis['pickup'].dt.dayofweek
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()
taxis['fin_semana'] = taxis['dia_num'] >= 5
taxis['fecha'] = taxis['pickup'].dt.date

# Comparar solo las fechas para detectar viajes que terminan otro día.
taxis['cruza_medianoche'] = taxis['pickup'].dt.date != taxis['dropoff'].dt.date
print('Viajes que cruzan la medianoche:', taxis['cruza_medianoche'].sum())
print(taxis[['pickup', 'dropoff', 'hora', 'dia_nombre',
             'fin_semana', 'fecha', 'cruza_medianoche']].head())

Viajes que cruzan la medianoche: 50
               pickup             dropoff  hora dia_nombre  fin_semana  \
0 2019-03-23 20:21:09 2019-03-23 20:27:24    20   Saturday        True   
1 2019-03-04 16:11:55 2019-03-04 16:19:00    16     Monday       False   
2 2019-03-27 17:53:01 2019-03-27 18:00:25    17  Wednesday       False   
3 2019-03-10 01:23:59 2019-03-10 01:49:51     1     Sunday        True   
4 2019-03-30 13:27:42 2019-03-30 13:37:14    13   Saturday        True   

        fecha  cruza_medianoche  
0  2019-03-23             False  
1  2019-03-04             False  
2  2019-03-27             False  
3  2019-03-10             False  
4  2019-03-30             False  


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [21]:
# La mediana reduce el efecto de los viajes de duración extrema.
por_hora = taxis.groupby('hora').agg(
    viajes=('hora', 'size'),
    duracion_mediana=('duracion_min', 'median')
)
print(por_hora)
print('\nHora con más viajes:', por_hora['viajes'].idxmax(),
      'con', por_hora['viajes'].max())
print('Hora con menos viajes:', por_hora['viajes'].idxmin(),
      'con', por_hora['viajes'].min())
print('Hora con mayor duración mediana:', por_hora['duracion_mediana'].idxmax(),
      'con', round(por_hora['duracion_mediana'].max(), 2), 'minutos')

      viajes  duracion_mediana
hora                          
0        205         10.733333
1        110          9.633333
2        101          9.183333
3         67          9.583333
4         57         10.500000
5         51          8.700000
6        142          7.833333
7        221          9.583333
8        313         11.316667
9        319         11.616667
10       325         11.883333
11       295         12.033333
12       334         12.775000
13       316         10.558333
14       360         11.841667
15       330         12.058333
16       336         11.583333
17       388         11.975000
18       417         11.100000
19       406         10.650000
20       368          9.258333
21       355         11.083333
22       321          9.816667
23       296         10.575000

Hora con más viajes: 18 con 417
Hora con menos viajes: 5 con 51
Hora con mayor duración mediana: 12 con 12.78 minutos


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [22]:
# Agrupar por número y nombre para mantener el orden lunes a domingo.
por_dia = taxis.groupby(['dia_num', 'dia_nombre']).agg(
    viajes=('fecha', 'size'),
    duracion_mediana=('duracion_min', 'median'),
    tarifa_mediana=('fare', 'median'),
    fechas_distintas=('fecha', 'nunique')
).sort_index()

# Dividir entre las fechas presentes de cada día, no entre siete.
por_dia['viajes_por_fecha'] = por_dia['viajes'] / por_dia['fechas_distintas']
print(por_dia.round(2).to_string())

                    viajes  duracion_mediana  tarifa_mediana  fechas_distintas  viajes_por_fecha
dia_num dia_nombre                                                                              
0       Monday         708             10.54             9.5                 4            177.00
1       Tuesday        825             11.22             9.5                 4            206.25
2       Wednesday      966             11.29             9.5                 4            241.50
3       Thursday       905             11.77             9.5                 5            181.00
4       Friday        1115             11.15             9.5                 5            223.00
5       Saturday      1046             10.47             9.0                 5            209.20
6       Sunday         868              9.98             9.0                 5            173.60


### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

La hora con más viajes fue las 18:00, con 417 y la de menos fue las 05:00, con 51. La mayor duración mediana fue a las 12:00 (12.78 minutos), así que no coincide con la hora de más viajes y puede relacionarse con el tráfico o el tipo de recorrido, aunque los datos no permiten saber la causa exacta.

Al revisar los días de marzo, el viernes tuvo el mayor total de viajes: 1,115. Al dividir por las fechas presentes, el miércoles queda arriba con 241.50 viajes por fecha, frente a 223.00 el viernes. Se necesita hacer esa división porque algunos días aparecen cinco veces y otros cuatro. En la tabla original el jueves aparece cinco veces porque incluye el viaje del 28 de febrero, asi que también filtré ese viaje para que la tabla incluyera solamente inicios de marzo.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [23]:
# Al aplicar limites de 2.4, una fila puede cumplir más
# de una condición, por eso el total se obtiene con una sola máscara OR.
fuera_periodo = ~pickup_en_marzo
corta_o_negativa = taxis['duracion_min'] < 1
muy_larga = taxis['duracion_min'] > 180
muy_rapida = taxis['velocidad_mph'] > 60
excluir = fuera_periodo | corta_o_negativa | muy_larga | muy_rapida

taxis_limpio = taxis.loc[~excluir].copy()
print('Fuera de marzo:', fuera_periodo.sum())
print('Duración menor a 1 minuto (incluye cero y negativas):', corta_o_negativa.sum())
print('Duración mayor a 180 minutos:', muy_larga.sum())
print('Velocidad mayor a 60 mph:', muy_rapida.sum())
print('Viajes excluidos sin contarlos dos veces:', excluir.sum())
print('Antes:', len(taxis), '| Después:', len(taxis_limpio))

# Verificar que el resultado sí cumple el criterio.
print('Fuera de marzo en el dataset limpio:',
      (~taxis_limpio['pickup'].between(inicio_marzo, inicio_abril, inclusive='left')).sum())
print('Duraciones fuera del intervalo [1, 180] minutos:',
      (~taxis_limpio['duracion_min'].between(1, 180)).sum())
print('Velocidades mayores a 60 mph:',
      (taxis_limpio['velocidad_mph'] > 60).sum())
print('Velocidades faltantes:', taxis_limpio['velocidad_mph'].isna().sum())

Fuera de marzo: 1
Duración menor a 1 minuto (incluye cero y negativas): 72
Duración mayor a 180 minutos: 0
Velocidad mayor a 60 mph: 11
Viajes excluidos sin contarlos dos veces: 74
Antes: 6433 | Después: 6359
Fuera de marzo en el dataset limpio: 0
Duraciones fuera del intervalo [1, 180] minutos: 0
Velocidades mayores a 60 mph: 0
Velocidades faltantes: 0


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [24]:
# Comparar los inicios de 07:00 a 09:59, separando días hábiles y fin de semana.
por_la_manana = taxis_limpio['hora'].between(7, 9)
comparacion = taxis_limpio.loc[por_la_manana].groupby('fin_semana').agg(
    viajes=('hora', 'size'),
    duracion_mediana_min=('duracion_min', 'median'),
    velocidad_mediana_mph=('velocidad_mph', 'median'),
    fechas_distintas=('fecha', 'nunique')
)
comparacion['viajes_por_fecha'] = comparacion['viajes'] / comparacion['fechas_distintas']
comparacion.index = comparacion.index.map({False: 'Días hábiles', True: 'Fin de semana'})
print(comparacion.round(2).to_string())

               viajes  duracion_mediana_min  velocidad_mediana_mph  fechas_distintas  viajes_por_fecha
fin_semana                                                                                            
Días hábiles      683                 12.03                   9.10                21             32.52
Fin de semana     161                  7.77                  13.47                10             16.10


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

Convertí ambas columnas con el formato %Y-%m-%d %H:%M:%S y no aparecieron NaT en ninguna de las 6,433 filas ya que las horas son locales de Nueva York. La revisión del 10 de marzo encontró 185 inicios, ninguno en la hora inexistente de las 02:00, tz_localize('America/New_York', nonexistent='NaT') tampoco detectó horas inexistentes en pickup.
Para estudiar duración y velocidad dejé solo los viajes que inician en marzo, duran entre 1 y 180 minutos y tienen velocidad promedio de 60 mph o menos. Había 1 inicio fuera de marzo, 72 duraciones menores a un minuto, 0 mayores a 180 minutos y 11 velocidades mayores a 60 mph. Algunas condiciones se superponen: se excluyeron 74 viajes únicos y quedaron 6,359, tambien conservo los viajes que cruzan al 1 de abril si iniciaron en marzo.
Entre las 07:00 y las 09:59, en días hábiles hubo 683 viajes, con mediana de 12.03 minutos y 9.10 mph, en fin de semana hubo 161, con 7.77 minutos y 13.47 mph. Como hay 21 fechas hábiles y 10 de fin de semana, también comparé viajes por fecha: 32.52 frente a 16.10. La muestra no permite asegurar que la diferencia se deba solo al tráfico. Sin un identificador o información adicional tampoco puedo confirmar si todos los viajes cortos eran errores, el cálculo de duración se hizo con las horas locales tal como aparecen y convendría verificar viajes que crucen un cambio de horario en otros datasets.

---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1. El formato explícito %Y-%m-%d %H:%M:%S deja claro cómo se interpretan las fechas y permite comprobar que las 6,433 se convirtieron sin NaT. Con textos como 03/04/2019, puede ser confuso porque deja que pandas adivine y podria interpretarlo como 3 de abril o 4 de marzo sin avisar.

2. Si pickup estuviera en UTC y dropoff en hora local, la resta directa de los textos convertidos daría una duración equivocada, normalmente varias horas menor e incluso negativa. Se deberian revisar los rangos, las duraciones negativas y el cambio de horario, después localizaría cada columna en su zona correcta y las compararía en la misma zona antes de restarlas. Aquí hubo 0 duraciones negativas, pero eso no basta por sí solo para demostrar que ambas columnas tengan la misma zona.

3. Un modelo podría aprender de duraciones imposibles o velocidades extremas como si fueran viajes normales. En esta muestra hubo 6 duraciones cero, 72 menores a un minuto y 11 velocidades mayores a 60 mph, 10 de estas últimas se superponen con las duraciones cortas y esas filas afectarían los patrones que use el modelo, sobre todo si se le pide predecir la duración de un viaje habitual.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.